rfcs/20200205-standalone-keras-repository.md

```diff
@@ -0,0 +1,361 @@
+implementation details. If certain private features are extensively used,
+we might want to consider exposing them  as public low level API.
+
+This design is also aligned with the design for
+[Modular TensorFlow](https://github.com/tensorflow/community/blob/master/rfcs/20190305-modular-tensorflow.md), 
+which splits the TensorFlow project into smaller components that are not
+tightly coupled together.
+
+
+## Design Proposal
+
+### New location of the code
+
+GitHub: the code will live at [keras-team/keras](https://github.com/keras-team/keras), 
+joining the other Keras SIG projects and replacing the current external Keras 
+codebase. `tf.Keras` will also replace Keras on PyPI.
+
+Also considered: `tensorflow/keras`.
+
+Pros:
+1. Under the umbrella of Keras SIG, which hosts all other Keras related projects
+like keras-application, KerasTuner etc.
+1. Lots of existing followers on keras-team, who may not be easily migrated to 
+TF project.
+1. Can't easily delete keras project, which already have tons of stars and 
+incoming reference links. Continued existence of external Keras code will create
+confusion ("why is there tensorflow/keras AND keras-team/keras?").
+
+Cons:
+1. The repo isn't under the same organization as tensorflow, which makes it hard
+to manage issues/PRs and references across the organization.
+1. Existing issue/PR under the same org can be transferred easily, but not cross the different org. See [here](https://help.github.com/en/github/managing-your-work-on-github/transferring-an-issue-to-another-repository).
+
+### Source of Truth
+
+TensorFlow uses a Google-internal code repository as its source of truth. Every PR
+submitted though GitHub is converted to a Google-internal change first,
+submitted through the internal system, and then copied to GitHub as commits.
+At the same time, PR is marked as merged with the corresponding commit hash.
+
+Likewise, issue tracking and code review takes place through Google-internal tools.
+
+For Keras, since we are trying to promote community engagement, we hope to use 
+GitHub as source of truth. This will have the following implications:
+
+* We expect the majority of the code development/contribution from GitHub
+and the dev tools / tests / scripts should focus on the GitHub development use
+case. See below for more details.
+* Keras CI/presubmit build for the GitHub repo should target a stable PIP 
+version of tensorflow package as dependency. It could either be a `tf-nightly`
+with explicit version, or a release candidate version, or a stable version.
+Depend on a floating `tf-nightly` could cause CI build to be instable, which has
+been observed in other repository 
+[like tf-addons](https://github.com/tensorflow/addons/pull/912).
+* The Keras code will be mirrored to a Google-internal code repository via
+Google-internal tools within a very short time window after each change.
+The Google-internal CI tests will run on HEAD for both Keras and TF code.
```

Comment:
I don't believe it's a good software practice because of sudden build breaks. But it's Google's problem. I just hope it won't come back to bite the community later (like with bazel) with a justification along the lines of "we do that internally so the community has to do the same too.".

Reply:
Ack, I think the both Francois and others on team is aware of this issue, and as we are pushing towards OSS first, we don't expect this to come back to bite the community.